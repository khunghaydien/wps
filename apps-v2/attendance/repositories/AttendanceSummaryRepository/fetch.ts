import Api from '@apps/commons/api';

import {
  BaseAttendanceSummary,
  convert,
} from '../models/BaseAttendanceSummary';
import {
  AttendanceSummary as DomainAttendanceSummary,
  IAttendanceSummaryRepository,
  Status,
} from '@attendance/domain/models/AttendanceSummary';

export type Response = BaseAttendanceSummary & {
  summaryName: string;
  status: Status;
  hasCalculatedAbsence: boolean;
};

const $convert = ({
  hasCalculatedAbsence,
  status,
  summaryName,
  ...response
}: Response): DomainAttendanceSummary => ({
  ...convert(response),
  name: summaryName,
  status,
  hasCalculatedAbsence,
});

const fetch: IAttendanceSummaryRepository['fetch'] = async (param) => {
  const response: Response = await Api.invoke({
    path: '/att/summary/get',
    param: {
      empId: param?.employeeId,
      targetDate: param?.targetDate,
    },
  });
  return $convert(response);
};

export default fetch;
