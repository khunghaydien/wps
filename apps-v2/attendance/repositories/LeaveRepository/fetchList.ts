import Api from '@apps/commons/api';

import * as DomainLeave from '@attendance/domain/models/Leave';
import * as DomainLeaveRange from '@attendance/domain/models/LeaveRange';
import * as DomainLeaveType from '@attendance/domain/models/LeaveType';

import createMapByCode from '@attendance/libraries/utils/Array/createMapByCode';

export type Leave = {
  name: string;
  code: string;
  type: DomainLeaveType.LeaveType;
  ranges: DomainLeaveRange.LeaveRange[];
  details: DomainLeave.Detail[];
  daysLeft: number | null;
  hoursLeft: number | null;
  timeLeaveDaysLeft: number | null;
  timeLeaveHoursLeft: number | null;
  requireReason: boolean;
};

export type Response = {
  leaves: Leave[];
};

const convert = (param: Leave): DomainLeave.Leave => ({
  ...DomainLeave.defaultValue,
  ...param,
  ranges: DomainLeaveRange.sort(param.ranges) || [],
  details: createMapByCode<DomainLeave.Detail>(
    param.details?.map((value) => ({
      ...value,
      ranges: DomainLeaveRange.sort(value.ranges) || [],
    }))
  ),
});

export default (async (param) =>
  Api.invoke({
    path: '/att/daily-leave/list',
    param: {
      targetDate: param.targetDate || null,
      empId: param.employeeId || null,
      ignoredId: param.ignoredId || null,
    },
  }).then(({ leaves }: Response) =>
    createMapByCode(leaves.map((value) => convert(value)))
  )) as DomainLeave.ILeaveRepository['fetchList'];
