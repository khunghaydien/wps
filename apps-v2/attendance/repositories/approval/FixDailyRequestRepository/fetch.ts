import Api from '@apps/commons/api';

import {
  AttendanceSummary,
  convert,
} from '@attendance/repositories/approval/models/AttendanceSummary';

import {
  FixDailyRequest,
  IFixDailyRequestRepository,
  Status,
} from '@attendance/domain/models/approval/FixDailyRequest';

export type Response = {
  id: string;
  status: Status;
  photoUrl: string;
  departmentName: string;
  delegatedEmployeeName: string;
  // comment: string;
  requestDate: string;
  targetDate: string;
} & AttendanceSummary;

const $convert = ({
  photoUrl: employeePhotoUrl,
  delegatedEmployeeName,
  ...request
}: Response): FixDailyRequest => {
  const convertedRequest = convert(request);
  const targetRecord = convertedRequest.records.find(
    (record) => record.recordDate === request.targetDate
  );
  return {
    id: request.id,
    requestDate: request.requestDate,
    targetDate: request.targetDate,
    submitter: {
      employee: {
        name: request.employeeName,
        code: request.employeeCode,
        photoUrl: employeePhotoUrl,
        department: {
          name: request.departmentName,
        },
      },
      delegator: {
        employee: {
          name: delegatedEmployeeName,
        },
      },
    },
    targetRecord,
    ...request,
    ...convertedRequest,
  };
};

const fetch: IFixDailyRequestRepository['fetch'] = async (requestId) => {
  const response: Response = await Api.invoke({
    path: '/att/request/fix-daily/get',
    param: {
      requestId,
    },
  });
  return $convert(response);
};

export default fetch;
