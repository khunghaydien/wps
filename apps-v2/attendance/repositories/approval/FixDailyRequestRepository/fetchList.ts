import Api from '@apps/commons/api';

import {
  AttendanceSummary,
  createDividedSummaries,
  createOwnerInfos,
  createRecord,
  createSummaries,
  createWorkingType,
} from '@attendance/repositories/approval/models/AttendanceSummary';

import {
  FixDailyRequest,
  IFixDailyRequestRepository,
  Status,
} from '@attendance/domain/models/approval/FixDailyRequest';

export type Response = {
  requestList: {
    id: string;
    status: Status;
    employeeInfoList: {
      startDate: string;
      endDate: string;
      departmentName: string;
      workingTypeName: string;
    }[];
    employeeCode: string;
    employeeName: string;
    photoUrl: string;
    departmentName: string;
    delegatedEmployeeName: string;
    // comment: string;
    requestDate: string;
    targetDate: string;
    record: AttendanceSummary['records'][number];
    recordTotal: {
      restTime: number;
      realWorkTime: number;
      overTime: number;
      nightTime: number;
      lostTime: number;
      virtualWorkTime: number;
      holidayWorkTime: number;
    };
    useRestReason: boolean;
    summaries: AttendanceSummary['summaries'];
    dividedSummaries: AttendanceSummary['dividedSummaries'];
  }[];
};

const $convert = (response: Response): FixDailyRequest[] => {
  const { requestList } = response;
  return requestList.map(
    ({
      photoUrl: employeePhotoUrl,
      departmentName,
      delegatedEmployeeName,
      record,
      recordTotal,
      useRestReason,
      ...request
    }) => {
      const targetRecord = createRecord(record);
      return {
        id: request.id,
        requestDate: request.requestDate,
        targetDate: request.targetDate,
        startDate: null,
        endDate: null,
        submitter: {
          employee: {
            name: request.employeeName,
            code: request.employeeCode,
            photoUrl: employeePhotoUrl,
            department: {
              name: departmentName,
            },
          },
          delegator: {
            employee: {
              name: delegatedEmployeeName,
            },
          },
        },
        ownerInfos: createOwnerInfos(request),
        targetRecord,
        records: [targetRecord],
        recordTotal,
        useRestReason: false,
        summaries: createSummaries(request.summaries, { records: [record] }),
        dividedSummaries: createDividedSummaries(request.dividedSummaries, {
          records: [record],
        }),
        attention: {
          ineffectiveWorkingTime: 0,
          insufficientRestTime: 0,
        },
        workingType: createWorkingType({ records: [record], useRestReason }),
        dailyAllowanceRecords: null,
        dailyObjectiveEventLogRecords: null,
        dailyRestRecords: null,
        displayFieldLayout: null,
      };
    }
  );
};

const fetchList: IFixDailyRequestRepository['fetchList'] = async ({
  employeeId,
  approvalType,
}) => {
  const response: Response = await Api.invoke({
    path: '/att/request-list/fix-daily/get',
    param: {
      empId: employeeId,
      approvalType,
    },
  });
  return $convert(response);
};

export default fetchList;
