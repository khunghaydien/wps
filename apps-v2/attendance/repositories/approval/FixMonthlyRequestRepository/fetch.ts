import Api from '@apps/commons/api';

import adapter from '@apps/repositories/adapters';

import { AttendanceSummary, convert } from '../models/AttendanceSummary';
import { ApprovalHistoryList } from '@apps/domain/models/approval/request/History';
import {
  FixMonthlyRequest,
  IFixMonthlyRequestRepository,
  Status,
} from '@attendance/domain/models/approval/FixMonthlyRequest';

export type Response = {
  id: string;
  status: Status;
  employeePhotoUrl: string;
  delegatedEmployeeName: string;
  comment: string;
} & AttendanceSummary &
  ApprovalHistoryList;

const $convert = ({
  employeePhotoUrl,
  delegatedEmployeeName,
  ...response
}: Response): FixMonthlyRequest => {
  return {
    ...response,
    ...convert(response),
    submitter: {
      employee: {
        code: response.employeeCode,
        name: response.employeeName,
        photoUrl: employeePhotoUrl,
        department: {
          name: '',
        },
      },
      delegator: {
        employee: {
          name: delegatedEmployeeName,
        },
      },
    },
  };
};

const fetch: IFixMonthlyRequestRepository['fetch'] = async (id) => {
  const response = await Api.invoke({
    path: '/att/request/monthly/get',
    param: {
      requestId: id,
    },
  });
  return adapter.fromRemote(response, [$convert]);
};

export default fetch;
