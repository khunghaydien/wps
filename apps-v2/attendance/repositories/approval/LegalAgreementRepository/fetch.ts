import Api from '@apps/commons/api';

import adapter from '@apps/repositories/adapters';

import { ApprovalHistoryList } from '@apps/domain/models/approval/request/History';
import * as DomainLegalAgreementRequest from '@attendance/domain/models/approval/LegalAgreementRequest';
import { WorkSystem as LegalAgreementWorkSystem } from '@attendance/domain/models/LegalAgreementOvertime';
import {
  CODE,
  Code,
} from '@attendance/domain/models/LegalAgreementRequestType';

type Common<TCode extends Code> = {
  id: string;
  status: DomainLegalAgreementRequest.Status;
  employeeName: string;
  employeePhotoUrl: string;
  delegatedEmployeeName: string;
  typeLabel: string;
  type: TCode;
};

type MonthlyRequest = Common<typeof CODE.MONTHLY> & {
  monthlyOvertimeHours: number | null;
  monthlyOvertimeHours1MoAgo: number | null;
  monthlyOvertimeHours2MoAgo: number | null;
  specialMonthlyOvertimeHours: number | null;
  specialMonthlyOvertimeHours1MoAgo: number | null;
  specialMonthlyOvertimeHours2MoAgo: number | null;
  overtimeHoursLimit: number | null;
  extensionCount: number | null;
  extensionCountLimit: number | null;
  changedOvertimeHoursLimit: number | null;
  requestableOvertimeHoursLimit: number | null;
  reason: string | null;
  measure: string | null;
};

type YearlyRequest = Common<typeof CODE.YEARLY> & {
  yearlyOvertimeHours: number | null;
  yearlyOvertimeHours1YearAgo: number | null;
  specialYearlyOvertimeHours: number | null;
  specialYearlyOvertimeHours1YearAgo: number | null;
  overtimeHoursLimit: number | null;
  changedOvertimeHoursLimit: number | null;
  requestableOvertimeHoursLimit: number | null;
  reason: string | null;
  measure: string | null;
};

type Request = MonthlyRequest | YearlyRequest;

export type Response = {
  request: Request;
  originalRequest: Request | null;
  legalAgreementWorkSystem: LegalAgreementWorkSystem;
} & ApprovalHistoryList;

const convertSubmitter = (request: Request) => ({
  employee: {
    name: request.employeeName,
    code: '',
    photoUrl: request.employeePhotoUrl,
    department: {
      name: '',
    },
  },
  delegator: {
    employee: {
      name: request.delegatedEmployeeName,
    },
  },
});

const convertCommon = (
  request: Request
): DomainLegalAgreementRequest.BaseRequest<Code> => ({
  id: request.id,
  status: request.status,
  submitter: convertSubmitter(request),
  type: request.type,
  typeLabel: request.typeLabel,
});

const convertMonthlyRequest = (
  request: MonthlyRequest
): DomainLegalAgreementRequest.MonthlyRequest => ({
  ...(convertCommon(request) as DomainLegalAgreementRequest.BaseRequest<
    typeof CODE.MONTHLY
  >),
  monthlyOvertimeHours: request.monthlyOvertimeHours,
  monthlyOvertimeHours1MoAgo: request.monthlyOvertimeHours1MoAgo,
  monthlyOvertimeHours2MoAgo: request.monthlyOvertimeHours2MoAgo,
  specialMonthlyOvertimeHours: request.specialMonthlyOvertimeHours,
  specialMonthlyOvertimeHours1MoAgo: request.specialMonthlyOvertimeHours1MoAgo,
  specialMonthlyOvertimeHours2MoAgo: request.specialMonthlyOvertimeHours2MoAgo,
  overtimeHoursLimit: request.overtimeHoursLimit,
  extensionCount: request.extensionCount,
  extensionCountLimit: request.extensionCountLimit,
  changedOvertimeHoursLimit: request.changedOvertimeHoursLimit,
  requestableOvertimeHoursLimit: request.requestableOvertimeHoursLimit,
  reason: request.reason,
  measure: request.measure,
});

const convertYearlyRequest = (
  request: YearlyRequest
): DomainLegalAgreementRequest.YearlyRequest => ({
  ...(convertCommon(request) as DomainLegalAgreementRequest.BaseRequest<
    typeof CODE.YEARLY
  >),
  yearlyOvertimeHours: request.yearlyOvertimeHours,
  yearlyOvertimeHours1YearAgo: request.yearlyOvertimeHours1YearAgo,
  specialYearlyOvertimeHours: request.specialYearlyOvertimeHours,
  specialYearlyOvertimeHours1YearAgo:
    request.specialYearlyOvertimeHours1YearAgo,
  overtimeHoursLimit: request.overtimeHoursLimit,
  changedOvertimeHoursLimit: request.changedOvertimeHoursLimit,
  requestableOvertimeHoursLimit: request.requestableOvertimeHoursLimit,
  reason: request.reason,
  measure: request.measure,
});

const convert = (
  response: Response
): DomainLegalAgreementRequest.LegalAgreementRequest => {
  const { request, originalRequest } = response;
  switch (request.type) {
    case CODE.MONTHLY:
      return {
        request: convertMonthlyRequest(request),
        originalRequest: originalRequest
          ? convertMonthlyRequest(originalRequest as MonthlyRequest)
          : null,
        historyList: response.historyList,
        legalAgreementWorkSystem: response.legalAgreementWorkSystem,
      } as DomainLegalAgreementRequest.MonthlyLegalAgreementRequest;
    case CODE.YEARLY:
      return {
        request: convertYearlyRequest(request),
        originalRequest: originalRequest
          ? convertYearlyRequest(originalRequest as YearlyRequest)
          : null,
        historyList: response.historyList,
        legalAgreementWorkSystem: response.legalAgreementWorkSystem,
      } as DomainLegalAgreementRequest.YearlyLegalAgreementRequest;
    default:
      return null;
  }
};

const fetch: DomainLegalAgreementRequest.ILegalAgreementRequestRepository['fetch'] =
  async (id) => {
    const response = await Api.invoke({
      path: '/att/request/legal-agreement/get',
      param: {
        requestId: id,
      },
    });
    return adapter.fromRemote(response, [convert]);
  };

export default fetch;
