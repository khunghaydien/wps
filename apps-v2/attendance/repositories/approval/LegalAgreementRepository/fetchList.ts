import Api from '@apps/commons/api';

import {
  ILegalAgreementRequestRepository,
  Status,
} from '@apps/attendance/domain/models/approval/LegalAgreementRequest';
import { Code } from '@attendance/domain/models/LegalAgreementRequestType';

type Response = {
  requestList: {
    id: string;
    employeeName: string;
    delegatedEmployeeName: string | null;
    photoUrl: string;
    departmentName: string | null;
    approverName: string;
    approverPhotoUrl: string;
    approverDepartmentName: string | null;
    requestDate: string;
    targetMonth: string;
    requestType: Code;
    requestStatus: Status;
    originalRequestStatus: Status;
  }[];
};

const fetchList: ILegalAgreementRequestRepository['fetchList'] = async (
  approvalType,
  selectId
) => {
  const response: Response = await Api.invoke({
    path: '/att/request-list/legal-agreement/get',
    param: { approvalType, selectId },
  });
  return response.requestList;
};

export default fetchList;
