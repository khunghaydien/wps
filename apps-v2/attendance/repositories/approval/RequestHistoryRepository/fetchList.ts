import Api from '@commons/api';

import adapter from '@apps/repositories/adapters';

import { IRequestHistoryRepository } from '@attendance/domain/models/approval/RequestHistory';

export default ((requestId) => {
  return Api.invoke({
    path: '/approval/request/history/get',
    param: adapter.toRemote({
      requestId,
    }),
  }).then(({ historyList }) =>
    adapter.fromRemote(historyList.map((history) => ({ ...history })))
  );
}) as IRequestHistoryRepository['fetchList'];
