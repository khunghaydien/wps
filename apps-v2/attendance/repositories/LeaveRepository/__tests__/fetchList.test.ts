import Api from '../../../../../__tests__/mocks/ApiMock';
import fetch from '../fetchList';
import { defaultValue } from './mocks/fetchList';

it('should return leaves', async () => {
  // Arrange
  const MockApi = Api.invoke as jest.Mock;
  MockApi.mockResolvedValueOnce(defaultValue);

  // Act
  const dailyRestList = await fetch({
    employeeId: 'employeeId',
    targetDate: '2022-07-01',
    ignoredId: '0001',
  });

  // Assert
  expect(dailyRestList).toMatchSnapshot();
  expect(MockApi).toBeCalledTimes(1);
  expect(MockApi).toBeCalledWith({
    path: '/att/daily-leave/list',
    param: {
      empId: 'employeeId',
      targetDate: '2022-07-01',
      ignoredId: '0001',
    },
  });
});
