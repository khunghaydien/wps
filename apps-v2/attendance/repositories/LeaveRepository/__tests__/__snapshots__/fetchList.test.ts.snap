// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return leaves 1`] = `
Map {
  "TEST" => Object {
    "code": "TEST",
    "daysLeft": 14,
    "details": Map {
      "TEST_DETAIL" => Object {
        "code": "TEST_DETAIL",
        "name": "Test Detail",
        "ranges": Array [
          "AM",
          "PM",
        ],
      },
    },
    "hoursLeft": 6,
    "name": "Test",
    "ranges": Array [
      "AM",
      "PM",
    ],
    "requireReason": true,
    "timeLeaveDaysLeft": 3,
    "timeLeaveHoursLeft": 2,
    "type": "Annual",
  },
}
`;
